### ISSUE TITLE
Explore page performance issues

### ISSUE THREAD

**Issue by ZeRego** [lightdash/lightdash#2997]
### Description

Explore page feels unresponsive since it might take seconds to update! 
Any action from:
- select a field from the sidebar
- add a filter
- switch from chart to big number or table 
- change any config in the chart/table

### To Reproduce

![Aug-18-2022 17-30-24](https://user-images.githubusercontent.com/9117144/185447352-bb1c40f7-cbcd-4f25-adef-0a94363096d7.gif)

Possible causes/solutions: 
- we are using `lazy: false` for the configuration popups. This means all the hooks and logic in those components are "active" and possibly causing multiple state changes in the Visualization provider. This can be easily be fixed and tested.
- Visualization provider is handling simultaneously all the chart types , big number and table hooks which are constantly recalculating configuration, etc. We should find a way to avoid this. Perhaps the hooks live in the specific type component and then save the state in the provider, once the user changes type the component calculates the state and replaces the state in the provider.
- Poor use of arrays when we should be using key/value maps. This is a recurrent issue when dealing with the explore items causing us to loop over data unnecessarily. I think the explore provider should provide the itemMap and the selectedItemMap. But I don't think this would have such a big impact compared with the previous points that may cause multiple re-renders. 
- Plotting data will most likely have to be moved to the server.

### App Version

0.222.0

**Comment by ZeRego** [comment]
When we open the chart section the page crashes when handling large amounts of data ( eg: 10000 ) and multiple columns

**Comment by ZeRego** [comment]
Main problem is explain in [this blog post ](https://www.nielskrijger.com/posts/2021-02-16/use-reducer-and-use-context/) 
**tldr:**  useContext is fairly dumb, it simply re-renders all subscribed components each time something changes. 

In the explore page we have a context that keeps track of the page state ( what sections are collapsed, selected fields, results, query, etc ) so each time we open/close a section it rerenders most of the page :/ 
The same problem happens with the visualization provider/context.

**Comment by rephus** [comment]
last time i looked briefly into it I discovered we load all chart configs on every change, and tableConfig has some heavy functions. Maybe we should not do unecessary calculations if the chart type is different...

https://github.com/lightdash/lightdash/blob/main/packages/frontend/src/components/LightdashVisualization/VisualizationProvider.tsx#L109

Note: We use sometimes that config on some other places, eg: we reuse columnOrder from tableConfig on tableVisualization (even if the chart type is not table) . For this we might want to find an alternative solution that doesn't depend on the config being uptodate.

**Comment by owlas** [comment]
Some performance data. This is from me clicking between the fields in the table config panel. 

### tl;dr

* Rendering is happening twice (nearly x2 load times)
* The hotkeys provider is triggering which reloads the entire page (twice, see above)
* Rendering the table is quite expensive
* onBlur will trigger a prop update to table config, triggering all context clients to rerender (see Jose's comment above)

### details of interaction

1. (16ms) click table config -> focus cell > only cell renders
2. (75ms) click another field -> viz provider re-renders from `tableConfig.state` change on column properties -> triggered from onBlur -> rerender table config panel (10ms) -> simple table (61ms)
3. (150ms) explore provider rerenders from `reducer` -> probably triggered by table update -> rerenders sidebar (15ms), viz (73ms), results table (58ms)
4. (118ms) hotkeys provider rerenders from `reducer` -> **investigate** -> rerender navbar (4ms), sidebar (13ms), viz (64ms), results table (28ms)
5. (39ms) hotkeys provider rerenders again -> **investigate** -> rerender things that depend on hotkeys (fast), table provider from `.useReactTable.state` (25ms)
6. (48ms) viz provider from `tableConfig.state` change on column properties -> rerenders config panel (5ms), viz (41ms)
7. (85ms) explore provider fires again (see 3)
8. (95ms) hotkeys provider fires again (see 4)
9. hotkeys provedir fires (see 5)

**Comment by TuringLovesDeathMetal** [comment]
#3083 might be related

**Comment by owlas** [comment]
The entire explore is getting fetched repeated when dbt is refreshing, it seems this hook is pretty trigger happy: https://www.loom.com/share/61cc322007a841fbbb88e590c0627d71

**Comment by rephus** [comment]
> Rendering the table is quite expensive - a lot of this from the cell context menu

This should fix it https://github.com/lightdash/lightdash/pull/3099

**Comment by IrakliJani** [comment]
following up to @rephus comment

here's the loom explaining the changes behind the https://github.com/lightdash/lightdash/pull/3099 pr
https://www.loom.com/share/5d60defe4d9f473ab17cc8716fd7e594

**Comment by ZeRego** [comment]
The 2 big performance issues I discovered so far that trigger most of the page to re-render:

**First problem:**
**useContext** doesn't let you subscribe to a part of the context value (or some memoized selector) without fully re-rendering.
They have an [open Issue](https://github.com/facebook/react/issues/14110) and there is a [proposal under development/review](https://github.com/reactjs/rfcs/pull/119).
So in [this PR](https://github.com/lightdash/lightdash/pull/3108) I'm using a [library](https://github.com/dai-shi/use-context-selector) that polyfills the proposal so once it's introduced in react we can easily migrate.

**Second problem:**
We update the url search params on each explore state change. 
React router components and hooks ( eg: useParams ) trigger an update causing most of the app the rerender even though the main route and params didn't change.
Haven't researched many solutions for this, but here is a list of potential fixes:
- move the useParams hook to the app provider and memoize the results
- stop saving the entire explore state in the url. Have users click a button to copy the url to that explore ( similar to what we have for saved charts ).

**Comment by ZeRego** [comment]
Closing this ticket as we already tackled quite a few performance issues in this page. 
I created a ticket specifically for the visualisation performance issue https://github.com/lightdash/lightdash/issues/3164
